autopush-common/src/reliability.rs

```diff
@@ -239,26 +292,66 @@ impl PushReliability {
                     // Write the expiration only if the state is non-terminal. Otherwise we run the risk of
                     // messages reporting a false "expired" state even if they were "successful".
                     let key = ExpiryKey {
-                        id: id.to_owned(),
+                        id: id.to_string(),
                         state: new.to_owned(),
                     }
                     .to_string();
-                    let _ = pipe.zadd(EXPIRY, &key, expr.unwrap_or_default());
+                    pipe.zadd(EXPIRY, &key, expr.unwrap_or_default());
                     trace!("🔍 internal record result: {:?}", key);
                 }
+                trace!("🔍 upping {:?}", &new);
+                // Bump up the new state count, and set the state key's state if it still exists.
+                pipe.hincr(COUNTS, new.to_string(), 1);
+                let options = redis::SetOptions::default()
+                    .conditional_set(redis::ExistenceCheck::XX)
+                    .get(true);
+                pipe.set_options(&state_key, new.to_string(), options);
                 // `exec_query` returns `RedisResult<()>`.
                 // `query_async` returns `RedisResult<Option<redis::Value>, RedisError>`.
                 // We really don't care about the returned result here, but transaction
                 // retries if we return Ok(None), so we run the exec and return
                 // a nonce `Some` value.
                 // The turbo-fish is a fallback for edition 2024
-                pipe.query_async::<()>(conn).await.inspect_err(|e| {
-                    warn!("🔍 Redis internal storage error: {:?}", e);
-                })?;
-                Ok(Some(redis::Value::Okay))
+                let result = pipe.query_async::<redis::Value>(conn).await?;
+                // The last element returned from the command is the result of the pipeline.
+                // The last of which should be the result of the `SET` command, which has `GET`
+                // set. This should either return the prior value or `Ok` if things worked, else
+                // it should return `nil`.
+                let rcheck = result
+                    .as_sequence()
+                    .unwrap_or_default()
+                    .last()
+                    .map(|v| {
+                        v.as_sequence()
```

Comment:
There is a lot of `default` and `unwrap_or()`. If we need to debug, it's probably hard to figure out from this code. Can we add comments or add explicit defaults?

Reply:
This is a case of being overly paranoid.
Redis will return `nil` if a lock fails, or if a `SET XX GET` operation fails because the value did not previously exist.
Technically, the latter should never happen because we're setting a transaction lock on both the `expiry` table and the `state.{ID}` entry, the value that we're doing the SET/GET.
I had collapsed this into a single operation partly because I didn't want to draw out the conditionals, but I can understand that it impacted readability.
I've reworked things to be clearer and more verbose, and also just recorded an `error!()` instead of rerunning the transaction, which should make things more readable.